### PR TITLE
Don't allow clients to send tombstones that reference the same room

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

**Comment by DaniellMesquita** [comment]
Will it automatically fix rooms that are already redirected to themselves?
I have one (LOL) here: https://matrix.to/#/!pQhDfsljSUQxaPNpur:matrix.org?via=matrix.org&via=amorgan.xyz&via=t2bot.io

**Comment by anoadragon453** [comment]
> Will it automatically fix rooms that are already redirected to themselves?
> I have one (LOL) here: https://matrix.to/#/!pQhDfsljSUQxaPNpur:matrix.org?via=matrix.org&via=amorgan.xyz&via=t2bot.io

I'm afraid not. This only prevents it from happening in the future.

To fix that case, just send another tombstone event in the room? Riot won't let you, but you can use `curl`.

File: changelog.d/5801.misc [hunk 1]
```diff
@@ -0,0 +1 @@
+Don't allow clients to send tombstone events that reference the room its sent in.
```
Comment (anoadragon453) [inline changelog.d/5801.misc]:
```suggestion
Don't allow clients to send tombstone events that reference the room it's sent in.
```

File: synapse/events/validator.py [hunk 2]
```diff
@@ -106,6 +106,13 @@ def validate_builder(self, event):
             if event.content["membership"] not in Membership.LIST:
                 raise SynapseError(400, "Invalid membership key")
 
+        elif event.type == EventTypes.Tombstone:
+            if "replacement_room" not in event.content:
+                raise SynapseError(400, "Content has no replacement_room key")
+
+            if event.content["replacement_room"] == event.room_id:
+                raise SynapseError(400, "Tombstone cannot reference itself")
```
Comment (anoadragon453) [inline synapse/events/validator.py]:
```suggestion
                raise SynapseError(400, "Tombstone cannot reference the room it was sent in")
```

**Pull request merged**